src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/CalculationTransactionalService.kt

```diff
@@ -39,18 +42,117 @@ class CalculationTransactionalService(
   private val prisonService: PrisonService,
   private val domainEventPublisher: DomainEventPublisher,
   private val prisonApiDataMapper: PrisonApiDataMapper,
-  private val calculationService: CalculationService
+  private val calculationService: CalculationService,
+  private val bookingService: BookingService,
+  private val validationService: ValidationService,
 ) {
 
   fun getCurrentAuthentication(): AuthAwareAuthenticationToken =
     SecurityContextHolder.getContext().authentication as AuthAwareAuthenticationToken?
       ?: throw IllegalStateException("User is not authenticated")
 
+  /*
+   * There are 4 stages of full validation:
+   * 1. validate user input data and the raw sentence and offence data from NOMIS
+   * 2. Validate the transformation of the raw Booking (NOMIS offence data transformed into a Booking object)
+   * 3. Run the calculation and catch any errors thrown by the calculation algorithm
+   * 4. Validate the post calculation Booking (The Booking is transformed during the calculation). e.g. Consecutive sentences (aggregates)
+   */
+  fun fullValidation(prisonerId: String, calculationUserInputs: CalculationUserInputs): List<ValidationMessage> {
+    val sourceData = prisonService.getPrisonApiSourceData(prisonerId)
+    var messages = validationService.validateSourceData(sourceData) // Validation stage 1 of 4
+    if (messages.isNotEmpty()) return messages
+
+    val booking = bookingService.getBooking(sourceData, calculationUserInputs)
+    messages = validationService.validateBookingBeforeCalculation(booking) // Validation stage 2 of 4
+    if (messages.isNotEmpty()) return messages
+    try {
+      val bookingAfterCalculation = calculationService.calculate(booking) // Validation stage 3 of 4
+      validationService.validateBookingAfterCalculation(bookingAfterCalculation) // Validation stage 4 of 4
```

Comment:
So changing the validation to work in these 4 stages means users may see error messages, go and fix those problems, try and calculate again and then see different error messages. I think stage 1 and 2 should definitely be combined. You may need to let Scott and Cherky know some of the rules you've implemented will come after others

Reply:
as discussed - will take this discussion to Cherky and do any actions in the next PR (validate of booking currently only affects FTR)